runtime/runtime-params-estimator/src/replay/fold_db_ops.rs

```diff
@@ -0,0 +1,263 @@
+    /// Print the DB operations that are at indentation 0, not associated with a span.
+    print_top_level: bool,
+    /// Only show data for a specific smart contract, specified by account id.
+    /// Currently only applies within receipts.
+    account_filter: Option<String>,
+    /// Skip all input lines that have deeper indentation.
+    skip_indent: Option<usize>,
+    /// Optionally collect and print detailed statistics for cache hits and misses.
+    track_caches: bool,
+    /// Keeps track of current block.
+    block_hash: Option<String>,
+    /// Stack of states.
+    states: Vec<State>,
+}
+
+#[derive(Default)]
+struct State {
+    /// The indent at which this state started
+    indent: usize,
+    /// Keeps track of operations per DB column.
+    ops_cols: BTreeMap<String, BTreeMap<String, usize>>,
+    /// Optionally collect and print detailed statistics for cache hits and misses.
+    cache_stats: Option<CacheStats>,
+}
+
+impl FoldDbOps {
+    pub(super) fn new() -> Self {
+        Self {
+            fold_anchors: vec![],
+            printed_fields: vec![],
+            print_top_level: true,
+            account_filter: None,
+            track_caches: false,
+            states: vec![State::default()],
+            block_hash: None,
+            skip_indent: None,
+        }
+    }
+
+    /// Pre-set that folds on chunks.
+    pub(super) fn chunks(self) -> Self {
+        self.fold("apply_transactions").print_field("receiver").print_field("shard_id")
+    }
+
+    /// Pre-set that folds on receipts.
+    pub(super) fn receipts(self) -> Self {
+        self.fold("process_receipt")
+            .print_field("receiver")
+            .print_field("receipt_id")
+            .fold("process_transaction")
+            .print_field("receipt_id")
+            .print_top_level(false)
+    }
+
+    pub(super) fn fold(mut self, anchor: impl Into<String>) -> Self {
+        self.fold_anchors.push(anchor.into());
+        self
+    }
+
+    pub(super) fn print_field(mut self, field: impl Into<String>) -> Self {
+        self.printed_fields.push(field.into());
+        self
+    }
+
+    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
+        self.print_top_level = yes;
+        self
+    }
+
+    pub(super) fn with_cache_stats(mut self) -> Self {
+        self.track_caches = true;
+        self.state().cache_stats = Some(CacheStats::default());
+        self
+    }
+
+    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
+        self.account_filter = account;
+        self
+    }
+
+    fn state(&mut self) -> &mut State {
+        self.states.last_mut().expect(EMPTY_STATE)
+    }
+
+    fn push_state(&mut self, indent: usize) {
+        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
+        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
+        self.states.push(new_state);
+    }
+
+    fn pop_state(&mut self) -> State {
```

Comment:
Is it intended that outer states don’t aggregate stats from inner states? I’m a bit confused why it is the case, because I expected the top-level state to aggregate everything.

Reply:
Yes that is intended. My thinking is this: A per-receipt output is mostly interested in the output for each receipt. But there is also some stuff that happens in between, which I called the top-level. Knowing how much is not associated with a receipt seems more interesting in this output than knowing the total.
Also, the total can still be retrieved by calling another command.

Does that make sense? Or would you say it makes more sense to use aggregate by default?

Reply:
It makes sense, thank you!